storage: TestStoreRangeMergeRHSLeaseExpiration failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/f5e3c29b2eed92868cf3d449575283e2e383f199

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1089030&tab=buildLog

```
I190111 06:33:02.098226 31113 gossip/gossip.go:391  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:35633" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I190111 06:33:02.098400 31400 gossip/client.go:128  [n2] started gossip client to 127.0.0.1:33709
I190111 06:33:02.100682 31113 storage/replica_command.go:349  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I190111 06:33:02.111477 31113 storage/store_snapshot.go:762  [s1,r1/1:{/Min-b}] sending preemptive snapshot 80d1dc9a at applied index 18
I190111 06:33:02.111651 31113 storage/store_snapshot.go:805  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 20, log entries: 8, rate-limit: 2.0 MiB/sec, 0.00s
I190111 06:33:02.114715 31624 storage/replica_raftstorage.go:809  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=80d1dc9a, encoded size=2633, 1 rocksdb batches, 8 log entries)
I190111 06:33:02.115565 31624 storage/replica_raftstorage.go:815  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190111 06:33:02.115940 31113 storage/replica_command.go:899  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I190111 06:33:02.117197 31113 storage/replica.go:3570  [s1,r1/1:{/Min-b},txn=26724b44] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190111 06:33:02.256402 31113 storage/store_snapshot.go:762  [s1,r2/1:{b-/Max}] sending preemptive snapshot edf651ca at applied index 12
I190111 06:33:02.256648 31113 storage/store_snapshot.go:805  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 0.00s
I190111 06:33:02.256973 31488 storage/replica_raftstorage.go:809  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=edf651ca, encoded size=7524, 1 rocksdb batches, 2 log entries)
I190111 06:33:02.257376 31488 storage/replica_raftstorage.go:815  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190111 06:33:02.257852 31113 storage/replica_command.go:899  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I190111 06:33:02.260382 31113 storage/replica.go:3570  [s1,r2/1:{b-/Max},txn=fc1bdb8f] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190111 06:33:02.316116 31643 storage/replica_command.go:492  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I190111 06:33:02.321623 31113 storage/client_test.go:1317  test clock advanced to: 1.800000125,0
I190111 06:33:02.334305 31498 storage/client_merge_test.go:1314  starting get 0
I190111 06:33:02.335355 31683 storage/client_merge_test.go:1314  starting get 1
I190111 06:33:02.336418 31499 storage/node_liveness.go:727  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I190111 06:33:02.336505 31684 storage/client_merge_test.go:1314  starting get 2
I190111 06:33:02.339798 31628 storage/client_merge_test.go:1314  starting get 3
I190111 06:33:02.340920 31629 storage/client_merge_test.go:1314  starting get 4
I190111 06:33:02.342232 31500 storage/client_merge_test.go:1314  starting get 5
I190111 06:33:02.343334 31501 storage/client_merge_test.go:1314  starting get 6
I190111 06:33:02.344437 31502 storage/client_merge_test.go:1314  starting get 7
I190111 06:33:02.345549 31672 storage/client_merge_test.go:1314  starting get 8
I190111 06:33:02.346654 31685 storage/client_merge_test.go:1314  starting get 9
I190111 06:33:02.413038 31512 storage/store.go:2658  [s2,r1/2:{/Min-b},txn=e1b0affd] removing replica r2/2
W190111 06:33:02.414080 31608 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (0.33s)
	client_merge_test.go:1343: expected RangeNotFound error from get during merge, but got <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/f5e3c29b2eed92868cf3d449575283e2e383f199

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1089307&tab=buildLog

```
W190111 06:54:46.663808 31397 gossip/gossip.go:1499  [n2] no incoming or outgoing connections
I190111 06:54:46.665088 31397 gossip/gossip.go:391  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:38653" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I190111 06:54:46.666659 31605 gossip/client.go:128  [n2] started gossip client to 127.0.0.1:43003
I190111 06:54:46.746517 31397 storage/replica_command.go:349  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I190111 06:54:46.764508 31397 storage/store_snapshot.go:762  [s1,r1/1:{/Min-b}] sending preemptive snapshot 309dc54f at applied index 18
I190111 06:54:46.765249 31397 storage/store_snapshot.go:805  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 20, log entries: 8, rate-limit: 2.0 MiB/sec, 0.00s
I190111 06:54:46.766718 31597 storage/replica_raftstorage.go:809  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=309dc54f, encoded size=2635, 1 rocksdb batches, 8 log entries)
I190111 06:54:46.769284 31597 storage/replica_raftstorage.go:815  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I190111 06:54:46.771263 31397 storage/replica_command.go:899  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I190111 06:54:46.777702 31397 storage/replica.go:3570  [s1,r1/1:{/Min-b},txn=3c6c04f4] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190111 06:54:46.930239 31397 storage/store_snapshot.go:762  [s1,r2/1:{b-/Max}] sending preemptive snapshot de3b8c52 at applied index 12
I190111 06:54:46.931059 31397 storage/store_snapshot.go:805  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 42, log entries: 2, rate-limit: 2.0 MiB/sec, 0.00s
I190111 06:54:46.931988 31609 storage/replica_raftstorage.go:809  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=de3b8c52, encoded size=7524, 1 rocksdb batches, 2 log entries)
I190111 06:54:46.933283 31609 storage/replica_raftstorage.go:815  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190111 06:54:46.935151 31397 storage/replica_command.go:899  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I190111 06:54:46.944317 31397 storage/replica.go:3570  [s1,r2/1:{b-/Max},txn=0e7b624c] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190111 06:54:47.053372 31490 storage/replica_command.go:492  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I190111 06:54:47.080847 31397 storage/client_test.go:1317  test clock advanced to: 1.800000125,0
I190111 06:54:47.831231 31614 storage/client_merge_test.go:1314  starting get 0
I190111 06:54:47.832423 31686 storage/client_merge_test.go:1314  starting get 1
I190111 06:54:47.833677 31687 storage/client_merge_test.go:1314  starting get 2
I190111 06:54:47.834912 31668 storage/client_merge_test.go:1314  starting get 3
I190111 06:54:47.836169 31616 storage/client_merge_test.go:1314  starting get 4
I190111 06:54:47.837467 31277 storage/client_merge_test.go:1314  starting get 5
I190111 06:54:47.838695 31617 storage/client_merge_test.go:1314  starting get 6
I190111 06:54:47.839816 31615 storage/node_liveness.go:727  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I190111 06:54:47.839928 31688 storage/client_merge_test.go:1314  starting get 7
I190111 06:54:47.841202 31646 storage/client_merge_test.go:1314  starting get 8
I190111 06:54:47.842427 31669 storage/client_merge_test.go:1314  starting get 9
I190111 06:54:47.971211 31560 storage/store.go:2658  [s2,r1/2:{/Min-b},txn=b9f7b751] removing replica r2/2
I190111 06:54:47.971282 31453 storage/store.go:2658  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.41s)
	client_merge_test.go:1343: expected RangeNotFound error from get during merge, but got <nil>

```

--- a/pkg/internal/client/sender.go
+++ b/pkg/internal/client/sender.go
@@ -170,13 +170,6 @@ type TxnSender interface {
 	// like the transaction that merges ranges together.
 	DisablePipelining() error
 
-	// EagerRecord instructs the transaction write its transaction record as soon as
-	// possible, instead of waiting for the transaction's first heartbeat or for the
-	// end of the transaction to write it.
-	//
-	// TODO(nvanbenschoten): Fix up flaky tests to allow us to get rid of this.
-	EagerRecord() error
-
 	// OrigTimestamp returns the transaction's starting timestamp.
 	// Note a transaction can be internally pushed forward in time before
 	// committing so this is not guaranteed to be the commit timestamp.
@@ -376,9 +369,6 @@ func (m *MockTransactionalSender) UpdateStateOnRemoteRetryableErr(
 
 // DisablePipelining is part of the client.TxnSender interface.
 func (m *MockTransactionalSender) DisablePipelining() error { return nil }
-
-// EagerRecord is part of the client.TxnSender interface.
-func (m *MockTransactionalSender) EagerRecord() error { return nil }
 
 // MockTxnSenderFactory is a TxnSenderFactory producing MockTxnSenders.
 type MockTxnSenderFactory struct {

--- a/pkg/internal/client/txn.go
+++ b/pkg/internal/client/txn.go
@@ -294,18 +294,6 @@ func (txn *Txn) DisablePipelining() error {
 	return txn.mu.sender.DisablePipelining()
 }
 
-// EagerRecord instructs the transaction write its transaction record as soon as
-// possible, instead of waiting for the transaction's first heartbeat or for the
-// end of the transaction to write it.
-//
-// EagerRecord must be called before any operations are performed on the
-// transaction.
-func (txn *Txn) EagerRecord() error {
-	txn.mu.Lock()
-	defer txn.mu.Unlock()
-	return txn.mu.sender.EagerRecord()
-}
-
 // NewBatch creates and returns a new empty batch object for use with the Txn.
 func (txn *Txn) NewBatch() *Batch {
 	return &Batch{txn: txn}

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -600,17 +600,6 @@ func (tc *TxnCoordSender) DisablePipelining() error {
 	return nil
 }
 
-// EagerRecord is part of the client.TxnSender interface.
-func (tc *TxnCoordSender) EagerRecord() error {
-	tc.mu.Lock()
-	defer tc.mu.Unlock()
-	if tc.mu.active {
-		return errors.Errorf("cannot request an eager transaction record write on a running transaction")
-	}
-	tc.interceptorAlloc.txnHeartbeater.eagerRecord = true
-	return nil
-}
-
 // commitReadOnlyTxnLocked "commits" a read-only txn. It is equivalent, but
 // cheaper than, sending an EndTransactionRequest. A read-only txn doesn't have
 // a transaction record, so there's no need to send any request to the server.

--- a/pkg/kv/txn_interceptor_heartbeater.go
+++ b/pkg/kv/txn_interceptor_heartbeater.go
@@ -86,10 +86,6 @@ type txnHeartbeater struct {
 	// is to notify the TxnCoordSender to shut itself down.
 	asyncAbortCallbackLocked func(context.Context)
 
-	// When set to true, the transaction will always send a BeginTxn request to
-	// lay down a transaction record as early as possible.
-	eagerRecord bool
-
 	// mu contains state protected by the TxnCoordSender's mutex.
 	mu struct {
 		sync.Locker
@@ -186,7 +182,7 @@ func (h *txnHeartbeater) SendLocked(
 			ba.Txn.Key = anchor
 		}
 
-		if h.eagerRecord || !h.st.Version.IsActive(cluster.VersionLazyTxnRecord) {
+		if !h.st.Version.IsActive(cluster.VersionLazyTxnRecord) {
 			addedBeginTxn = true
 
 			// Set the key in the begin transaction request to the txn's anchor key.

--- a/pkg/storage/batcheval/cmd_lease_request.go
+++ b/pkg/storage/batcheval/cmd_lease_request.go
@@ -82,7 +82,7 @@ func RequestLease(
 		// time here, which makes it not Equivalent() to the preceding lease for the
 		// same store.
 		//
-		// Note also that leastPostApply makes sure to update the timestamp cache in
+		// Note also that leasePostApply makes sure to update the timestamp cache in
 		// this case: even though the lease holder does not change, the the sequence
 		// number does and this triggers a low water mark bump.
 		//

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -1202,12 +1202,12 @@ func TestStoreRangeMergeRHSLeaseExpiration(t *testing.T) {
 	// read the meat of the test.
 
 	// Install a hook to control when the merge transaction commits.
-	mergeEndTxnReceived := make(chan struct{}, 10) // headroom in case the merge transaction retries
+	mergeEndTxnReceived := make(chan *roachpb.Transaction, 10) // headroom in case the merge transaction retries
 	finishMerge := make(chan struct{})
 	storeCfg.TestingKnobs.TestingRequestFilter = func(ba roachpb.BatchRequest) *roachpb.Error {
 		for _, r := range ba.Requests {
 			if et := r.GetEndTransaction(); et != nil && et.InternalCommitTrigger.GetMergeTrigger() != nil {
-				mergeEndTxnReceived <- struct{}{}
+				mergeEndTxnReceived <- ba.Txn
 				<-finishMerge
 			}
 		}
@@ -1261,13 +1261,24 @@ func TestStoreRangeMergeRHSLeaseExpiration(t *testing.T) {
 
 	// Wait for the merge transaction to send its EndTransaction request. It won't
 	// be able to complete just yet, thanks to the hook we installed above.
-	<-mergeEndTxnReceived
+	mergeTxn := <-mergeEndTxnReceived
 
 	// Now's our chance to move the lease on the RHS from the second store to the
 	// first. This isn't entirely straightforward. The replica on the second store
 	// is aware of the merge and is refusing all traffic, so we can't just send a
 	// TransferLease request. Instead, we need to expire the second store's lease,
 	// then acquire the lease on the first store.
+
+	// Before doing so, however, ensure that the merge transaction has written
+	// its transaction record so that it doesn't run into trouble with the low
+	// water mark of the new leaseholder's timestamp cache. This could result in
+	// the transaction being inadvertently aborted during its first attempt,
+	// which this test is not designed to handle. If the merge transaction did
+	// abort then the get requests could complete on r2 before the merge retried.
+	hb, hbH := heartbeatArgs(mergeTxn, mtc.clock.Now())
+	if _, pErr := client.SendWrappedWith(ctx, mtc.stores[0].TestSender(), hbH, hb); pErr != nil {
+		t.Fatal(pErr)
+	}
 
 	// Turn off liveness heartbeats on the second store, then advance the clock
 	// past the liveness expiration time. This expires all leases on all stores.

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -1881,11 +1881,7 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	db := mtc.dbs[0]
 	txnOld := client.NewTxn(ctx, db, 0 /* gatewayNodeID */, client.RootTxn)
 
-	// Perform a write with txnOld so that its timestamp gets set and so
-	// that it writes its transaction record.
-	if err := txnOld.EagerRecord(); err != nil {
-		t.Fatal(err)
-	}
+	// Perform a write with txnOld so that its timestamp gets set.
 	if _, err := txnOld.Inc(ctx, keyB, 3); err != nil {
 		t.Fatal(err)
 	}
@@ -1894,6 +1890,16 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	// transaction from refreshing when it hits the serializable error.
 	if _, err := txnOld.Get(ctx, keyC); err != nil {
 		t.Fatal(err)
+	}
+
+	// Ensure that the transaction sends its first hearbeat so that it creates
+	// its transaction record and doesn't run into trouble with the low water
+	// mark of the new leaseholder's timestamp cache. Amusingly, if the bug
+	// we're regression testing against here still existed, we would not have
+	// to do this.
+	hb, hbH := heartbeatArgs(txnOld.Serialize(), mtc.clock.Now())
+	if _, pErr := client.SendWrappedWith(ctx, mtc.stores[0].TestSender(), hbH, hb); pErr != nil {
+		t.Fatal(pErr)
 	}
 
 	// Another client comes along at a higher timestamp and reads. We should

--- a/pkg/storage/client_test.go
+++ b/pkg/storage/client_test.go
@@ -1395,6 +1395,17 @@ func truncateLogArgs(index uint64, rangeID roachpb.RangeID) *roachpb.TruncateLog
 	}
 }
 
+func heartbeatArgs(
+	txn *roachpb.Transaction, now hlc.Timestamp,
+) (*roachpb.HeartbeatTxnRequest, roachpb.Header) {
+	return &roachpb.HeartbeatTxnRequest{
+		RequestHeader: roachpb.RequestHeader{
+			Key: txn.Key,
+		},
+		Now: now,
+	}, roachpb.Header{Txn: txn}
+}
+
 func TestSortRangeDescByAge(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	var replicaDescs []roachpb.ReplicaDescriptor

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -395,14 +395,6 @@ func (r *Replica) AdminMerge(
 		// TODO(benesch): expose a proper API for preventing the fast path.
 		_ = txn.CommitTimestamp()
 
-		// Multiple merge tests rely on a transaction record being written as
-		// early as possible in the lifecycle of a transaction. They aren't
-		// prepared for an untimely lease transfer to cause a restart.
-		// TODO(nvanbenschoten): Remove this once the tests are fixed.
-		if err := txn.EagerRecord(); err != nil {
-			return err
-		}
-
 		// Pipelining might send QueryIntent requests to the RHS after the RHS has
 		// noticed the merge and started blocking all traffic. This causes the merge
 		// transaction to deadlock. Just turn pipelining off; the structure of the

